daemon/src/GHCSpecter/UI/ConcurReplica/Types.hs

```diff
@@ -4,75 +4,46 @@ module GHCSpecter.UI.ConcurReplica.Types
   ( -- * IHTML type
     IHTML (..),
 
-    -- * IHTML <-> HTML
+    -- * project
     project,
-    embed,
 
-    -- * block / unblock DOM
+    -- * block
     blockDOMUpdate,
     unblockDOMUpdate,
   )
 where
 
 import Concur.Core
-  ( SuspendF (..),
-    Widget (Widget, step),
+  ( Widget,
     mapView,
   )
-import Control.Monad.Free (hoistFree)
-import Control.ShiftMap (ShiftMap (..))
 import Replica.VDOM (HTML)
 
 -- | IHTML has additional tag about whether one wants to bypass DOM update.
 -- For example, onMouseMove events are fired too frequently, and most of the handling action
 -- is just to update internal state, not leading to DOM changes.
 -- With IHTML, we tag the HTML content as non-update and bypass expensive websocket diff update steps.
--- Left: no need for update, Right: need for update
-newtype IHTML = IHTML {unIHTML :: Either HTML HTML}
+data IHTML
+  = -- | update
+    Update HTML
+  | -- | no update
+    NoUpdate HTML
 
 instance Semigroup IHTML where
-  IHTML (Left e1) <> IHTML (Left e2) = IHTML (Left (e1 <> e2))
-  IHTML (Left e1) <> IHTML (Right e2) = IHTML (Right (e1 <> e2))
-  IHTML (Right e1) <> IHTML (Left e2) = IHTML (Right (e1 <> e2))
-  IHTML (Right e1) <> IHTML (Right e2) = IHTML (Right (e1 <> e2))
+  Update e1 <> Update e2 = Update (e1 <> e2)
+  Update e1 <> NoUpdate e2 = NoUpdate (e1 <> e2)
+  NoUpdate e1 <> Update e2 = NoUpdate (e1 <> e2)
+  NoUpdate e1 <> NoUpdate e2 = NoUpdate (e1 <> e2)
 
 instance Monoid IHTML where
-  mempty = IHTML (Left mempty)
+  mempty = Update []
```

Comment:
Note that the semantics from mempty is changed. by default, it is updating without any element. and NoUpdate is the upper bound of the HTML (i.e. once NoUpdate, then always NoUpdate)

Reply:
I should add this comment.

Reply:
added